### PR TITLE
#1130 Consider the right target type for object factory methods

### LINKED ISSUE #1130: Adding @TargetType to @ObjectFactory method causes it not to be used in mapping with @MappingTarget parameter
In the example below both createB() and createC() methods seems like valid object factory methods for creating BDTO and CDTO objects respectively. But in mergeA() method implementation only createB() is used. In this method CDTO is created using constructor. In other generated methods createC() is used correctly.
 
```java
@Mapper
public interface VerifyMappings {
	class AEntity {
		private BEntity b;
		private CEntity c;
		public BEntity getB() {
			return b;
		}
		public void setB(BEntity b) {
			this.b = b;
		}
		public CEntity getC() {
			return c;
		}
		public void setC(CEntity c) {
			this.c = c;
		}
	}
	class BEntity {}
	class CEntity {}
	class ADTO {
		private BDTO b;
		private CDTO c;
		public BDTO getB() {
			return b;
		}
		public void setB(BDTO b) {
			this.b = b;
		}
		public CDTO getC() {
			return c;
		}
		public void setC(CDTO c) {
			this.c = c;
		}
	}
	class BDTO {}
	class CDTO {}

	ADTO copyA(AEntity source);
	BDTO copyB(BEntity source);
	CDTO copyC(CEntity source);

	void mergeA(AEntity source, @MappingTarget ADTO target);
	void mergeB(BEntity source, @MappingTarget BDTO target);
	void mergeC(CEntity source, @MappingTarget CDTO target);

	@ObjectFactory
	default BDTO createB() {
		return new BDTO();
	}

	@ObjectFactory
	default CDTO createC(@TargetType Class<CDTO> clazz) {
		return new CDTO();
	}
}
```

### LINKED ISSUE #1131: Source parameter for @ObjectFactory method is not well defined
It seems that the source parameter to object factory method has different meaning depending if it is used from method with @MappingTarget annotation or not. For methods with @MappingTarget parent of object which is going to be mapped into newly created object is passed as source and in case of methods without @MappingTarget annotation the object to be mapped is passed itself.
This can be seen in the example below where copyB() method implementation passes BDTO instance as source and in mergeA() method implementation passes ADTO instance.

```java
@Mapper
public interface VerifyMappings {
	class AEntity {
		private BEntity b;
		public BEntity getB() {
			return b;
		}
		public void setB(BEntity b) {
			this.b = b;
		}
	}
	class BEntity {}
	class ADTO {
		private BDTO b;
		public BDTO getB() {
			return b;
		}
		public void setB(BDTO b) {
			this.b = b;
		}
	}
	class BDTO {}

	ADTO copyA(AEntity source);
	BDTO copyB(BEntity source);

	void mergeA(AEntity source, @MappingTarget ADTO target);
	void mergeB(BEntity source, @MappingTarget BDTO target);

	@ObjectFactory
	default BDTO createB(Object source) {
		return new BDTO();
	}
}
```

### PR DESCRIPTION
**Pull request by agudian**
Fixes #1130.

### CODE REVIEW

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1131/Issue1131Test.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public class Issue1131Test {
+    @Test
+    public void factoryMethodWithTargetTypeInUpdateMethods() {
+        AEntity aEntity = new AEntity();
+        aEntity.setB( new BEntity() );
+
+        ADto aDto = new ADto();
+        Mappers.getMapper( Issue1131Mapper.class ).mergeA( aEntity, aDto );
+
+        assertThat( aDto.getB() ).isNotNull();
```
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1131/Issue1131Test.java]: should you also check the text or is that overkill?

**Comment by sjaakd** [comment]
Look ok to me.. !

**Comment by agudian** [comment]
Oh sorry guy, I was mistaken here - this is about #1130 - I'll rename the tests and the change the commit message tonight.
The question about #1134 remains to be discussed.

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1131/Issue1131Test.java (hunk 1 above)
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1131/Issue1131Test.java]: I'll add one... 👍

**Comment by agudian** [comment]
retest this, please.

**Comment by agudian** [comment]
Dear Jenkins, retest this, please!

**Comment by agudian** [comment]
Thanks for the reviews, Sjaak and Filip! 👍

**Pull request merged**